### PR TITLE
Chore: refactor docblock parsing into a more specific `DocBlock` value object with an associated registry

### PR DESCRIPTION
**Pull request by Ocramius**

### CODE REVIEW

File: src/Annotation/Registry.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Yes, this is a singleton registry. The reason why this can actually be a singleton
+ * without having kittens die is that reflection information (and therefore docblock
+ * information) is static within a single process.
+ *
+ * @internal This class is part of PHPUnit internals, an not intended
+ *           for downstream usage
+ *
+ * @TODO test me (srsly!)
```
Comment (Ocramius) [inline src/Annotation/Registry.php]: Yeap.

File: src/Annotation/DocBlock.php [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+                try {
+                    $versionConstraintParser = new VersionConstraintParser;
+
+                    $requires[$matches['name'] . '_constraint']        = [
+                        'constraint' => $versionConstraintParser->parse(\trim($matches['constraint'])),
+                    ];
+                    $recordedOffsets[$matches['name'] . '_constraint'] = $offset;
+                } catch (\PharIo\Version\Exception $e) {
+                    /* @TODO this catch is currently not valid, see https://github.com/phar-io/version/issues/16 */
```
Comment (Ocramius) [inline src/Annotation/DocBlock.php]: See https://github.com/phar-io/version/issues/16

**Comment by sebastianbergmann** [comment]
Looks like we found a bug in @muglug's Psalm:
```
Notice: Undefined property: DOMComment::$tagName in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php on line 84

Fatal error: Uncaught Error: Call to undefined method DOMComment::getAttribute() in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php:85
Stack trace:
#0 phar:///github/workspace/tools/psalm/src/psalm.php(407): _HumbugBox0ece64d7e152\Psalm\ErrorBaseline::read(Object(_HumbugBox0ece64d7e152\Psalm\Internal\Provider\FileProvider), '.psalm/baseline...')
#1 phar:///github/workspace/tools/psalm/psalm(5): require_once('phar:///github/...')
#2 /github/workspace/tools/psalm(14): require('phar:///github/...')
#3 {main}
  thrown in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php on line 85
```
https://github.com/sebastianbergmann/phpunit/pull/3836/checks?check_run_id=215062082

**Comment by Ocramius** [comment]
Possibly `baseline.xml` containing a comment:

```
    <!-- see https://github.com/phar-io/version/issues/16 -->
```

**Pull request merged**